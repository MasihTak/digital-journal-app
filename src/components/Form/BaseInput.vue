<script setup>
import {onMounted} from "vue";

const emit = defineEmits(["update:modelValue"]);

/*----------- Props -----------*/
const props = defineProps({
  label: {
    type: String,
    required: true
  },
  name: {
    type: String,
    required: true
  },
  id: {
    type: String,
    required: false
  },
  placeholder: {
    type: String,
    required: false
  },
  maxlength: {
    type: String,
    required: false
  },
  modelValue: {
    type: String,
    default: ""
  },
  error: {
    type: String,
    default: ""
  }
});

/*----------- Data -----------*/
/**
 * @desc if the id attribute omitted use the same value as the name.
 */
let identity = props.id ? props.id : props.name;

/*----------- onMounted -----------*/
/**
 * @desc Emit an update to the modelValue prop when the component is mounted.
 */
onMounted(() => {
  emit("update:modelValue", props.modelValue);
});
</script>

<template>
  <label :for="identity"
         class="d-block"
  >{{ label }}</label>
  <input
    :id="identity"
    :name="name"
    :maxlength="maxlength"
    :placeholder="placeholder"
    class="form-control"
    v-bind="$attrs"
    :value="modelValue"
    @input="$emit('update:modelValue', $event.target.value)"
  >
  <p v-if="error"
     class="invalid-feedback"
  >
    {{ error }}
  </p>
</template>
