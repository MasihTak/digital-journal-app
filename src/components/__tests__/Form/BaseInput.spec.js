import {describe, it, expect} from 'vitest';
import {mount} from '@vue/test-utils';
import BaseInput from "@/components/Form/BaseInput.vue";

describe('BaseInput.vue', () => {
    it('matches snapshot', () => {
        const label = 'Test Label'
        const name = 'Test Name'
        const id = 'Test ID'
        const placeholder = 'Test Placeholder'
        const maxlength = '10'
        const modelValue = 'Test Value'
        const error = 'Test Error'

        const wrapper = mount(BaseInput, {
            props: {
                label,
                name,
                id,
                placeholder,
                maxlength,
                modelValue,
                error
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('emits update:modelValue event on input', async () => {
        const wrapper = mount(BaseInput, {
            props: {
                label: 'Test Label',
                name: 'Test Name',
                modelValue: 'Test Value'
            }
        })

        await wrapper.find('input').setValue('New Value')

        // Check that the event has been emitted more than once
        expect(wrapper.emitted('update:modelValue').length).toBeGreaterThan(1)
        // Check that the last emitted value is 'New Value'
        expect(wrapper.emitted('update:modelValue').pop()).toMatchSnapshot()
    })
})
