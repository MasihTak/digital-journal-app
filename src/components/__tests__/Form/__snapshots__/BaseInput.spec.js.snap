// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInput.vue > emits update:modelValue event on input 1`] = `
[
  "New Value",
]
`;

exports[`BaseInput.vue > matches snapshot 1`] = `
"<label for="Test ID" class="d-block">Test Label</label>
<input id="Test ID" name="Test Name" maxlength="10" placeholder="Test Placeholder" class="form-control">
<p class="invalid-feedback">Test Error</p>"
`;
